### PR TITLE
SOLR-15244 Switch File to Path

### PR DESCRIPTION
**Pull request by madrob**
https://issues.apache.org/jira/browse/SOLR-15244

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/CloudUtil.java [hunk 1]
```diff
@@ -125,7 +125,7 @@ public static boolean replicaExists(ClusterState clusterState, String collection
   public static String unifiedResourcePath(SolrResourceLoader loader) {
     return (loader instanceof ZkSolrResourceLoader) ?
             ((ZkSolrResourceLoader) loader).getConfigSetZkPath() + "/" :
-            loader.getConfigDir() + File.separator;
+            loader.getConfigPath().toString() + File.separator;
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/cloud/CloudUtil.java:128]: Referencing File.separator seems wrong here; just use '/'.  toString() is automatic as well.

File: solr/core/src/java/org/apache/solr/cloud/SolrZkServer.java [hunk 2]
```diff
@@ -176,29 +176,24 @@ public void stop() {
...
             + " file is missing");
       }
 
-      Properties cfg = new Properties();
-      FileInputStream in = new FileInputStream(configFile);
-      try {
-        cfg.load(new InputStreamReader(in, StandardCharsets.UTF_8));
-      } finally {
-        in.close();
+      try (Reader reader = Files.newBufferedReader(configPath)) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/cloud/SolrZkServer.java:192]: love it

File: solr/core/src/java/org/apache/solr/core/MMapDirectoryFactory.java [hunk 3]
```diff
@@ -61,8 +60,7 @@ public void init(NamedList args) {
 
   @Override
   protected Directory create(String path, LockFactory lockFactory, DirContext dirContext) throws IOException {
-    // we pass NoLockFactory, because the real lock factory is set later by injectLockFactory:
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/MMapDirectoryFactory.java:64]: was this comment obsolete?

File: solr/core/src/java/org/apache/solr/core/SolrResourceLoader.java [hunk 4]
```diff
@@ -915,27 +921,22 @@ private PackageListeningClassLoader createSchemaLoader() {
...
-          final String msg = "Can't create managed schema directory " + parentDir.getAbsolutePath();
-          log.error(msg);
-          throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, msg);
-        }
-      }
-      try (OutputStream out = new FileOutputStream(confFile);) {
-        out.write(content);
+      Path parentDir = confFile.getParent();
+      if (!Files.isDirectory(parentDir)) {
+        Files.createDirectories(parentDir);
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/SolrResourceLoader.java:928]: according to javadocs, createDirectories doesn't throw if parentDir already exists so you needn't conditionally check.

File: solr/core/src/java/org/apache/solr/core/snapshots/SolrSnapshotsTool.java [hunk 5]
```diff
@@ -183,26 +180,29 @@ public void describeSnapshot(String collectionName, String snapshotName) {
...
+  /**
+   * @param pathPrefix optional
+   */
+  public Map<String, List<String>> getIndexFilesPathForSnapshot(String collectionName,  String snapshotName, String pathPrefix)
       throws SolrServerException, IOException {
     Map<String, List<String>> result = new HashMap<>();
 
     Collection<CollectionSnapshotMetaData> snaps = listCollectionSnapshots(collectionName);
-    Optional<CollectionSnapshotMetaData> meta = Optional.empty();
+    CollectionSnapshotMetaData meta = null;
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/snapshots/SolrSnapshotsTool.java:191]: I'm not a fan of Optional either

File: solr/core/src/java/org/apache/solr/core/snapshots/SolrSnapshotsTool.java [hunk 6]
```diff
@@ -226,10 +226,9 @@ public void describeSnapshot(String collectionName, String snapshotName) {
       }
 
       String indexDirPath = coreSnap.getIndexDirPath();
-      if (pathPrefix.isPresent()) {
+      if (pathPrefix != null) {
         // If the path prefix is specified, rebuild the path to the index directory.
-        Path t = new Path(coreSnap.getIndexDirPath());
-        indexDirPath = (new Path(pathPrefix.get(), t.toUri().getPath())).toString();
+        indexDirPath = new Path(pathPrefix, coreSnap.getIndexDirPath()).toString();
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/snapshots/SolrSnapshotsTool.java:231]: This is referring to Hadoop APIs! `org.apache.hadoop.fs.Path` Ugh!  Could it have been intentional and thus is there risk your change actually breaks something?

File: solr/core/src/java/org/apache/solr/core/snapshots/SolrSnapshotsTool.java [hunk 7]
```diff
@@ -244,20 +243,16 @@ public void describeSnapshot(String collectionName, String snapshotName) {
...
-        filesBuilder.append("\n");
-      }
-
-      String files = filesBuilder.toString().trim();
-      try (Writer w = new OutputStreamWriter(new FileOutputStream(new File(localFsPath, entry.getKey())), StandardCharsets.UTF_8)) {
-        w.write(files);
-      }
+      // TODO: this used to trim - check if that's needed
+      // Using Paths.get instead of Path.of because of conflict with o.a.hadoop.fs.Path
+      Files.write(Paths.get(localFsPath, entry.getKey()), entry.getValue());
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/snapshots/SolrSnapshotsTool.java:255]: I learned something new today; thanks!

File: solr/core/src/java/org/apache/solr/filestore/DistribPackageStore.java [hunk 8]
```diff
@@ -582,29 +581,26 @@ private static String _getMetapath(String path) {
    */
   public static void _persistToFile(Path solrHome, String path, ByteBuffer data, ByteBuffer meta) throws IOException {
     Path realpath = _getRealPath(path, solrHome);
-    File file = realpath.toFile();
-    File parent = file.getParentFile();
-    if (!parent.exists()) {
-      parent.mkdirs();
+    Path parent = realpath.getParent();
+    if (!Files.exists(parent)) {
+      Files.createDirectories(realpath.getParent());
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/filestore/DistribPackageStore.java:586]: again; probably don't need to check that it exists first

File: solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java [hunk 9]
```diff
@@ -1521,7 +1520,7 @@ public DirectoryFileStream(SolrParams solrParams) {
       
       sOffset = params.get(OFFSET);
       sLen = params.get(LEN);
-      compress = params.get(COMPRESSION);
+      compress = Boolean.parseBoolean(params.get(COMPRESSION));
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java:1523]: this snuck in here but okay

File: solr/test-framework/src/java/org/apache/solr/core/MockFSDirectoryFactory.java [hunk 10]
```diff
@@ -65,7 +64,7 @@ private Directory reduce(Directory dir) {
     if (dir instanceof NRTCachingDirectory) {
       cdir = ((NRTCachingDirectory)dir).getDelegate();
     }
-    if (cdir instanceof TrackingDirectoryWrapper) {
+    if (dir instanceof TrackingDirectoryWrapper) {
```
Comment (dsmiley) [inline solr/test-framework/src/java/org/apache/solr/core/MockFSDirectoryFactory.java:67]: uh... is this change right?

**Review by dsmiley — CHANGES_REQUESTED** [review]

File: solr/core/src/java/org/apache/solr/core/snapshots/SolrSnapshotsTool.java (hunk 6 above)
Comment (madrob) [reply solr/core/src/java/org/apache/solr/core/snapshots/SolrSnapshotsTool.java:231]: Yea, this change is still fine, the old code is doing a bunch of extra conversions between Hadoop Path, URI, and String, and back and forth.

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStatesOps.java [hunk 11]
```diff
@@ -65,13 +65,7 @@ private void persist(List<PerReplicaStates.Operation> operations, String znode,
...
                     Op.delete(path, -1));
         }
-        try {
-            zkClient.multi(ops, true);
-        } catch (KeeperException e) {
-          log.error("Multi-op exception: {}", zkClient.getChildren(znode, null, true));
-          throw e;
-        }
-
+        zkClient.multi(ops, true);
```
Comment (madrob) [inline solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStatesOps.java:68]: this snuck in and shouldn't be here, will remove before next iteration.

File: solr/core/src/java/org/apache/solr/core/MMapDirectoryFactory.java (hunk 3 above)
Comment (madrob) [reply solr/core/src/java/org/apache/solr/core/MMapDirectoryFactory.java:64]: yes

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStatesOps.java [hunk 12]
```diff
@@ -65,7 +65,13 @@ private static void persist(List<PerReplicaStates.Operation> operations, String
                     Op.create(path, null, zkClient.getZkACLProvider().getACLsToAdd(path), CreateMode.PERSISTENT) :
                     Op.delete(path, -1));
         }
-        zkClient.multi(ops, true);
+        try {
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStatesOps.java]: This snuck in?

**Pull request merged**